### PR TITLE
Implemented message passing between FileWatcher and Vert.x

### LINKED ISSUE #332: [core/dispatcher/receiver][data-plane] Reconciler should live in its own verticle, detached from FileWatcher thread
**Describe the bug**
Today we run the reconciler within the `FileWatcher` running thread. This makes weird the listeners and more slippy to implement, because they don't have the same guarantees as running inside `Verticle` (like the single thread guarantee, that allows to use non thread-safe data structures).

 **Expected behavior**
`Reconciler` should run within its own verticle and receive file watch updates from `FileWatcher`

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Fixes #332

## Proposed Changes

- Introduces a bunch of classes to send/receive the contract using the event bus: `ContractMessageCodec`, `ContractPublisher`, `ResourcesReconcilerMessageHandler`
- `ConsumerDeployer` becomes a verticle that has to be deployed to run. It starts the message consumer and the reconciler internally.
- `HttpVerticle` in receiver changes a bit its role and becomes `ReceiverVerticle`. Now, more than being a mere http server starter, it knows about the receiver classes and starts the reconciler message consumer together with the http server. This aligns it with `ConsumerDeployer` usage and role and also keeps the reconciliation on the same verticle of the listener.
- Renamed `RequestReceiver` to proper `RequestMapper`

### CODE REVIEW

File: data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java [hunk 1]
```diff
@@ -189,20 +191,12 @@ public void execute(final Vertx vertx, final VertxTestContext context) throws Ex
...
 
-    // reconcile ingress/egress
-    reconciler.reconcile(Collections.singletonList(resource))
-      .onFailure(context::failNow)
-      .onSuccess(v -> waitReconciler.countDown());
-
-    waitReconciler.await(TIMEOUT, TimeUnit.SECONDS);
+    //TODO(slinkydeveloper) for testing purpose, we need to implement a way to propagate results of reconciliation
+    // Or should we use awaitability or similar?
+    Thread.sleep(10000);
```
Comment (slinkydeveloper) [inline data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java:199]: For this one I'm thinking to add (later, in another pr) http://www.awaitility.org/ and wait on the number of deployed verticles, wdyt @pierDipi ?

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 2]
```diff
@@ -44,33 +45,38 @@
...
+
     final var reconciler = ResourcesReconcilerImpl
       .builder()
       .watchEgress(consumerDeployer)
       .build();
 
     reconciler.reconcile(resources)
       .onSuccess(ignored -> context.verify(() -> {
-        assertThat(vertx.deploymentIDs()).hasSize(numEgresses);
+        assertThat(vertx.deploymentIDs()).hasSize(numEgresses + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:79]: Add a comment for the `+1` or change it `+ NUM_SYSTEM_VERTICLES` or something similar.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 3]
```diff
@@ -80,30 +86,34 @@ public void shouldAddResourceAndDeployVerticles(final Vertx vertx, final VertxTe
...
+
     final var reconciler = ResourcesReconcilerImpl
       .builder()
       .watchEgress(consumerDeployer)
       .build();
 
     reconciler.reconcile(resources)
       .onFailure(ignored -> context.verify(() -> {
-        assertThat(vertx.deploymentIDs()).hasSize(0);
+        assertThat(vertx.deploymentIDs()).hasSize(1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:116]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 4]
```diff
@@ -150,13 +164,13 @@ public void shouldStopVerticleWhenEgressDeleted(
       .onSuccess(ignored -> {
 
         context.verify(() -> {
-          assertThat(vertx.deploymentIDs()).hasSize(numEgressesOld);
+          assertThat(vertx.deploymentIDs()).hasSize(numEgressesOld + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:167]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 5]
```diff
@@ -150,13 +164,13 @@ public void shouldStopVerticleWhenEgressDeleted(
...
           checkpoints.flag();
         });
 
         reconciler.reconcile(resourcesNew)
           .onSuccess(ok -> context.verify(() -> {
-            assertThat(vertx.deploymentIDs()).hasSize(numEgressesNew);
+            assertThat(vertx.deploymentIDs()).hasSize(numEgressesNew + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:173]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 6]
```diff
@@ -215,13 +233,13 @@ public void shouldStopVerticlesWhenResourceDeleted(
     reconciler.reconcile(resourcesOld)
       .onSuccess(ignored -> {
         context.verify(() -> {
-          assertThat(vertx.deploymentIDs()).hasSize(numEgressesOld);
+          assertThat(vertx.deploymentIDs()).hasSize(numEgressesOld + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:236]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 7]
```diff
@@ -215,13 +233,13 @@ public void shouldStopVerticlesWhenResourceDeleted(
...
           checkpoints.flag();
         });
 
         reconciler.reconcile(resourcesNew)
           .onSuccess(ok -> context.verify(() -> {
-            assertThat(vertx.deploymentIDs()).hasSize(numEgressesNew);
+            assertThat(vertx.deploymentIDs()).hasSize(numEgressesNew + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:242]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 8]
```diff
@@ -294,21 +316,21 @@ public void shouldStopAndStartVerticlesWhenEgressDeletedAndReAdded(
       .onSuccess(ignored -> {
 
         context.verify(() -> {
-          assertThat(oldDeployments).hasSize(numEgressesOld);
+          assertThat(oldDeployments).hasSize(numEgressesOld + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:319]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 9]
```diff
@@ -294,21 +316,21 @@ public void shouldStopAndStartVerticlesWhenEgressDeletedAndReAdded(
...
           checkpoints.flag();
         });
 
         reconciler.reconcile(resourcesNew)
           .onSuccess(ok -> {
             context.verify(() -> {
-              assertThat(vertx.deploymentIDs()).hasSize(numEgressesNew);
+              assertThat(vertx.deploymentIDs()).hasSize(numEgressesNew + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:326]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 10]
```diff
@@ -294,21 +316,21 @@ public void shouldStopAndStartVerticlesWhenEgressDeletedAndReAdded(
...
               assertThat(vertx.deploymentIDs()).containsAll(oldDeployments);
               checkpoints.flag();
             });
 
             reconciler.reconcile(resourcesOld)
               .onSuccess(ok2 -> context.verify(() -> {
-                assertThat(oldDeployments).hasSize(numEgressesOld);
+                assertThat(oldDeployments).hasSize(numEgressesOld + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:333]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java [hunk 11]
```diff
@@ -364,12 +390,12 @@ public void shouldDoNothingWhenTheStateIsTheSame(
         final var deployments = vertx.deploymentIDs();
 
         context.verify(() -> {
-          assertThat(deployments).hasSize(numEgresses);
+          assertThat(deployments).hasSize(numEgresses + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/ConsumerDeployerVerticleTest.java:393]: Same here.

File: data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/integration/UnorderedConsumerTest.java [hunk 12]
```diff
@@ -130,7 +130,7 @@ public void testUnorderedConsumer(final Vertx vertx, final VertxTestContext cont
 
     Thread.sleep(6000); // reduce flakiness
 
-    assertThat(vertx.deploymentIDs()).hasSize(numEgresses);
+    assertThat(vertx.deploymentIDs()).hasSize(numEgresses + 1);
```
Comment (pierDipi) [inline data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/integration/UnorderedConsumerTest.java:133]: Same here.

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/ReceiverVerticle.java [hunk 13]
```diff
@@ -16,30 +16,42 @@
...
    *
+   * @param livenessPath
+   * @param readinessPath
    * @param httpServerOptions server options.
    * @param requestHandler    request handler.
    */
-  public HttpVerticle(
+  public ReceiverVerticle(
+    final String livenessPath,
+    final String readinessPath,
```
Comment (pierDipi) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/ReceiverVerticle.java:50]: Why should this verticle know about probes?

File: data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java (hunk 1 above)
Comment (pierDipi) [reply data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java:199]: SGTM

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/ReceiverVerticle.java (hunk 13 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/ReceiverVerticle.java:50]: I'd prefer to keep it agnostic and remove noise in this verticle.

Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/ReceiverVerticle.java:50]: Ok I have another idea to hide it, let me try it

Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/ReceiverVerticle.java]: Ok what about now?

**Comment by slinkydeveloper** [comment]
flaky not related to this test
/retest

**Comment by slinkydeveloper** [comment]
@pierDipi sacura error https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka-broker/346/pull-knative-sandbox-eventing-kafka-broker-integration-tests/1320790707404804096#1:build-log.txt%3A63390

**Comment by pierDipi** [comment]
From the sacura job logs seems like [not all events have been sent](https://00f74ba44b22de64f623f2b4cd198f86325bc4f369-apidata.googleusercontent.com/download/storage/v1/b/knative-prow/o/pr-logs%2Fpull%2Fknative-sandbox_eventing-kafka-broker%2F346%2Fpull-knative-sandbox-eventing-kafka-broker-integration-tests%2F1320790707404804096%2Fartifacts%2Fee010f1e-17b6-11eb-b2a3-aa6241aa3ab6%2Fsacura%2Fsacura%2Fsacura-c5gnc-sacura.log?jk=AFshE3U9naZ4tvJJFZ-eg6YfEENrmwKkRy8S6ccCyq4knZGJGcjZQD9yRxhT8ZJPvLQhD07YIPBk9buy9AFlK5j9be1Ed4elnsADapKuf6DgUD0RsBFRFboOFK2BafqFGA4z1PfUVl6b8aSOMSckfopmB5WZsdvjq6Z8_gNkYSfQB1xW2ThQ4KytgEsP7zCbjLL2Y7E3icb3wTvgkv9AMb-8yuZ3YUou3HhEyz0VVLLUy28h6MuaiU_skG6k0gMOeZFPi1vBVq7VG3zPxzsAxBQ9ZdIyECk-ry35kFNO8drfif1t7DMeZbFHuUxgSTmkAbu0lOM4n41GuNyVgcNRgJ1gvqNtwmuQ-cW8NRjZvLy0JGyzuZfwK3qV2K2Np4F--XOdK99A8J93z_CakBSnA1xxhSiloK1GtxzVKvJARos2AAdDrvXqgEoVo6rc1XfPTPcXMlSrJyHcwGlcdC53xJHyTKpskw5rZixmMP6ZkbJi-s-k4k14Y_gcmCCoK2icZwmS8v_aJ4PkDlCQqesxoDgrkwqC6tyGg0XkvGMEv7qpDVSDB_QSbkVpeKcdI-j8n5q7GQzw6p6Y0DS-WauSRvzrcQnHlJx-5VUIp_Vrd1sHdNiLnkj5PGxpw-grSYXaDzHeTxaOH0MScynzJ9vFSRmGKseMretFT8JBo0Ww8kZg0I-2tXhblvaN3zGXJgb1WH5Dz47hsIROAfxuvJxi-KJFycuN3GQyv6avaQoh9r7QrDF-IYJuWkUY2Qfg0RusBw6UKBrkpHi_7oD6nHp-J2K-AToDCEq2fcXHMlHdjdPTNIa6aT0Apb64w4I_qSUblvFllusI5cwXJ8cRO-GYFycWANRdmIq5kkznUhrSMnezrrYlXnZL7WNaQ6Du8nNSi9nKAnV1l5Oq9A6e0L0Zwb4qg2gMNwty-Me1pvhiXTagAqL2lVJsTAlM1Ok-uPYgFD_9HkKAPlYLqnvZBWLdFOE13s2s3I3vmuRYVijBHCqNSKgvoCcSWEJynZLQFi-2pTqcG21aJ-Dv-ABFL6Lg4Amh290A6-_Ujw9QK2brjsQLcqeBlztUnO4kcav7YGnvCULYWxpCwyPHB8K_9JFiYcd8BttS8VbGiNGmYKbDsGKxA4eIMr0nR6UBCuLccSmjz9hgcE_D7Rrom4cwU0opV4QFd_frJ3unxQ&isca=1):
```
"status_codes": {
   "202": 119951
  },
```
It should be `"202": 120k`, so we have been slower to receive and/or consume events.

File: data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/reconciler/impl/ResourcesReconcilerMessageHandler.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  @Override
+  public void handle(Message<Object> event) {
+    DataPlaneContract.Contract contract = (DataPlaneContract.Contract) event.body();
+    resourcesReconciler.reconcile(contract.getResourcesList())
+      .onComplete(result -> {
+        if (result.succeeded()) {
+          logger.info("reconciled objects {}", keyValue("contract", contract));
+        } else {
+          logger.error("failed to reconcile {}", keyValue("contract", contract), result.cause());
+        }
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/reconciler/impl/ResourcesReconcilerMessageHandler.java:36]: Be aware that onSuccess/onFailure are more costly in terms of allocations, you should not abuse them on the hot path. In this case, it's fine

File: data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java (hunk 1 above)
Comment (pierDipi) [reply data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java:199]: Can you open an issue?

**Review by pierDipi — APPROVED** [review]
/lgtm

File: data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java (hunk 1 above)
Comment (slinkydeveloper) [reply data-plane/tests/src/test/java/dev/knative/eventing/kafka/broker/tests/DataPlaneTest.java:199]: done https://github.com/knative-sandbox/eventing-kafka-broker/issues/351

**Pull request merged**